File: Rotated_sorted_array_search.py
##Written by Ashutosh Upadhye
###Time Complexity:O(Log2(n))
##Space Complexity:O(log2(n)) for recursive implementation

class search_sorted_array:

    def binary_search_sorted(self,array, low, high, target):
        if low > high: #Check whether the bounds passed to the binary search are proper or not.If low > high the algorithm terminates
            return -1

        if low <= high:
            mid = low + (high - low) // 2 #compute mid

            if array[mid] == target: # if mid is the target return mid index
                return mid
            if array[low] <= array[mid]: #Look if left subarray array is sorted
                if target >= array[low] and target <= array[mid]: # Look if key exists inside that subarray or not
                    return search_sorted_array.binary_search_sorted(self,array, low, mid-1, target) #Search the left subarray
                return search_sorted_array.binary_search_sorted(self,array, mid + 1, high, target) # Or search the right subarray
            if target >= array[mid] and target <= array[high]:#If left subarray isnt sorted, assume that right subarray is sorted
                    return search_sorted_array.binary_search_sorted(self,array, mid+1, high, target)  #Search the right subarray
            return search_sorted_array.binary_search_sorted(self,array, low, mid-1, target)#Search the left subarray

    def search(self, nums, target):
        low = 0
        high = len(nums)-1
        index_to_be_found = search_sorted_array.binary_search_sorted(self,nums, low, high, target)
        return index_to_be_found #Return the index to be found


    def main(self):
        array = [3, 4, 5, 6, 7, 8, 9, 10, 1, 2]
        target = 15
        target_found = search_sorted_array.search(self, array, target)
        print(target_found)


Object = search_sorted_array().main()

File: Rotated_Sorted_Array_search.cpp
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : logical

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l=0, h=nums.size()-1;
        int m;
        while(l<=h){
            m = l + (h-l)/2;
            if(nums[m]==target){
                return m;
            }
            else if(nums[m]>=nums[l]){
                if(target>=nums[l] && target<nums[m]){
                    h = m-1;
                }
                else{
                    l = m+1;
                }
            }
            else{
                if(target>nums[m] && target<=nums[h]){
                    l = m+1;
                }
                else{
                    h = m-1;
                }
            }
        }
        return -1;
    }
};
